src/BoundaryConditions/field_boundary_conditions.jl

```diff
@@ -83,5 +83,5 @@ end
   VVelocityBoundaryConditions(grid; kwargs...) = FieldBoundaryConditions(grid, field_type=:v; kwargs...)
   WVelocityBoundaryConditions(grid; kwargs...) = FieldBoundaryConditions(grid, field_type=:w; kwargs...)
      TracerBoundaryConditions(grid; kwargs...) = FieldBoundaryConditions(grid, field_type=:tracer; kwargs...)
-#   PressureBoundaryConditions(grid; kwargs...) = FieldBoundaryConditions(grid, field_type=:pressure, kwargs...)
-#DiffusivityBoundaryConditions(grid; kwargs...) = FieldBoundaryConditions(grid, field_type=:diffusivity, kwargs...)
+   PressureBoundaryConditions(grid; kwargs...) = FieldBoundaryConditions(grid, field_type=:pressure; kwargs...)
```

Comment:
what is the advantage of the "field_type" keyword? I think this is poor design because it will be harder to maintain and the code is not transparent. I would rather define these constructors here.

Reply:
The `field_type` is used to call e.g. `default_tracer_bc` or `default_velocity_bc` in an `@eval` statement.

Yeah I wasn't super sure of the best way to define these but I did not want to define separate explicit constructors for each six of them as the function signature for each would be pretty complicated
```
function FieldBoundaryConditions(grid::AbstractGrid; field_type,
                                 west=default_bc(grid, field_type, 1), east=default_bc(grid, field_type, 1),
                                 south=default_bc(grid, field_type, 2), north=default_bc(grid, field_type, 2),
                                 bottom=default_bc(grid, field_type, 3), top=default_bc(grid, field_type, 3))
```
so I added a `field_type` kwarg and was able to write a single `FieldBoundaryConditions` function that handled all six field types.

I don't think harder to maintain (there's less code and really only one function that does anything), and I'm not sure what is not transparent about this piece of code.

Happy to change it, just not sure what improvements to make.

Reply:
Here's what I propose:

1. We define constructors for boundary conditions that take `location`, `topology` , and optional specified pairs in each direction. The defaults are

```julia
DefaultBoundaryCondition(loc, ::Union{Periodic, Flat}) = PeriodicBC()

DefaultBoundaryCondition(Cell, ::Bounded) = NoFluxBC()

DefaultBoundaryCondition(Face, ::Bounded) = NoPenetrationBC()
```

2. `FieldBoundaryConditions` takes a location tuple (eg `(Face, Cell, Cell)`) as a required positional argument (rather than `field_type`), and then creates 3 pairs of `CoordinateBoundaryConditions` for each direction, for example, 

```julia
function FieldBoundaryConditions(grid, loc; 
    east = DefaultBoundaryCondition(topology(grid)[1], loc[1]),
    west = DefaultBoundaryCondition(topology(grid)[1], loc[1]),
    south = DefaultBoundaryCondition(topology(grid)[2], loc[2]),
    north = DefaultBoundaryCondition(topology(grid)[2], loc[2]),
    bottom = DefaultBoundaryCondition(topology(grid)[3], loc[3]),
    top = DefaultBoundaryCondition(topology(grid)[3], loc[3])
)

# validate the boundary conditions

    x = CoordinateBoundaryConditions(west, east)
    y = CoordinateBoundaryConditions(south, north)
    z = CoordinateBoundaryConditions(bottom, top)

    return FieldBoundaryConditions(x, y, z)
end
```

3. The convenience constructors call `FieldBoundaryConditions` with appropriate arguments; for example:

```julia
UVelocityBoundaryConditions(grid; user_defined_bcs...) = FieldBoundaryConditions(grid, (Face, Cell, Cell); user_defined_bcs...)
```

This will allow users to define fields with "default" boundary conditions for fields that are not of any `field_type`. For example, vertical vorticity is located at `(Face, Face, Cell)`. Users can still avoid having to remember "where" a field is on the grid by using our convenience constructors. I also think this may (?) shorten and simplify the code and reduce shenanigans (maybe not desirable since we are `Oceananigans`...?) We still want to perform validation so some shenanigans are inevitable.

Does that make sense? Any issues I'm not forseeing?

Reply:
Just another thought: we can make the signatures defensive by adding `arch` as a required positional argument to `DefaultBoundaryCondition` so we are ready for distributed fields with `InteriorBoundaryCondition()` ?

Reply:
Thanks for the suggestions, looks great! Just updated the code, I do think it's simpler now.

Reply:
> Just another thought: we can make the signatures defensive by adding `arch` as a required positional argument to `DefaultBoundaryCondition` so we are ready for distributed fields with `InteriorBoundaryCondition()` ?

Hmmm, I thought about it a bit but am a little hesitant as I'm not sure what a distributed field might want. What I do in https://github.com/climate-machine/Oceananigans.jl/pull/590 is "inject halo communication boundary conditions" after the user has specified any non-default BCs so `arch` isn't necessarily needed but of course it's still being designed. There are also no distributed fields, just distributed models (a wrapper around a model plus MPI stuff like rank connectivity).

Should be an easy enough change to make when/if we need it.

Reply:
Fair enough.